### ISSUE TITLE
Allow "undo" on revertable actions

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-admin#468]
# Acceptance Criteria

 * I have the option to undo any revertable actions in the same context as taking them
 * The option is only presented directly after taking the original action
 * Applies to actions on versioned records: Archive, unpublish, publish
 * Applies to pages, campaigns and other versioned records

# Notes

 * Does not apply to form information (handled natively through browser)
 * Could be implemented in success notification
 * Undo should be tracked in version history (to avoid records being live without having a record of it)

# Related

 * [On archiving, the archive menu indicates activity](https://github.com/silverstripe/silverstripe-admin/issues/452)
 * [Ability to revert/unrelease campaigns](https://github.com/silverstripe/silverstripe-campaign-admin/issues/43)

**Comment by dhensby** [comment]
Is this meant describe a similar behaviour to how Google Inbox (and other apps) allows you to "undo" a send (and other actions)?

**Comment by chillu** [comment]
No, since Google Inbox effectively delays carrying out an action that becomes irreversible once performed (sending an email). In our case we'd carry out the action immediately, but allow authors to carry out the reversal later (as long as the notification is visible). This is just a shortcut to using specialised section like the upcoming "global archive" to restore records.

**Comment by chillu** [comment]
This is a rabbit hole of client-side state updates. If you unpublished a page, it'll reflect the new status in the tree view. If you now undo that (re-publishing the page), the tree view needs to reflect that. This is a good example of the in-depth UI changes which are best left to a wider UI refactor.

If you tackle this, please spell out the exact cases which will receive undo, and research which UI elements need to update based on it.
